### PR TITLE
[Wasm]Adjust AppContext.BaseDirectory and enable Microsoft.Extensions tests 

### PR DESCRIPTION
**Pull request by steveisok**
Adjust AppContext.BaseDirectory to return "/" for wasm because GetEntryAssembly().Location returns an empty string.

This change also enables most of the Microsoft.Extensions.* tests with the exception of DependencyModel and Hosting.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/AppContext.UnixOrWindows.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace System
+{
+    public static partial class AppContext
+    {
+        private static string GetBaseDirectoryCore()
+        {
+            // Fallback path for hosts that do not set APP_CONTEXT_BASE_DIRECTORY explicitly
+            string? directory = Path.GetDirectoryName(Assembly.GetEntryAssembly()?.Location);
+            if (directory != null && !Path.EndsInDirectorySeparator(directory))
```
Comment (marek-safar) [inline src/libraries/System.Private.CoreLib/src/System/AppContext.UnixOrWindows.cs:16]: Do the null check only one and return Empty here

File: src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace System
+{
+    public static partial class AppContext
+    {
+        private static string GetBaseDirectoryCore()
+        {
+            // GetEntryAssembly().Location returns an empty string for wasm
+            // Until that can be easily changed, work around the problem here.
+            return "/";
```
Comment (marek-safar) [inline src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs:13]: Should we add explicit test for this?

Comment (steveisok) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs:13]: Good question.  I think it's acceptable to just rely on the tests that use this property to pass.

File: src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+// See the LICENSE file in the project root for more information.
+
+namespace System
+{
+    public static partial class AppContext
+    {
+        private static string GetBaseDirectoryCore()
+        {
+            // GetEntryAssembly().Location returns an empty string for wasm
+            // Until that can be easily changed, work around the problem here.
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs:12]: How is this working around `GetEntryAssembly().Location` returning empty string?

Would it make more sense for `BaseDirectory` to be empty string as well when there is not a file system?

Comment (steveisok) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs:12]: Perhaps the wording is incorrect.  AppContext.BaseDirectory is returning "", which is causing a bunch of tests to fail.  I feel returning "/" isn't any less wrong than "".  The benefit of "/" being that it works within other parts of libraries.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs:12]: I think that returning "/" is more wrong than "".

If we were optimizing for making maximum tests to pass, we would also change Assembly.Location to return `/<assemblyname>.dll` that would be even more wrong.

What are the tests that fail before this change and pass with this change?

Comment (steveisok) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs:12]: I’ll follow up with a list.

Since the emscripten vfs mounts at /, I think my change does make sense.
https://emscripten.org/docs/porting/files/file_systems_overview.html

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs:12]: Do the application binaries show up as files at  `/` ?

The docs for [AppContext.BaseDirectory](https://docs.microsoft.com/en-us/dotnet/api/system.appcontext.basedirectory) say: Gets the pathname of the base directory that the assembly resolver uses to probe for assemblies.

File: src/libraries/Microsoft.Extensions.DependencyModel/tests/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Xunit;
+
+[assembly: ActiveIssue("https://github.com/dotnet/runtime/issues/37669", TestPlatforms.Browser)]
```
Comment (jkotas) [inline src/libraries/Microsoft.Extensions.DependencyModel/tests/AssemblyInfo.cs:8]: This issue talks about crypto. Is it correct issue to reference here?

I would expect Microsoft.Extensions.DependencyModel tests to have no or few dependencies on crypto.

Comment (steveisok) [reply src/libraries/Microsoft.Extensions.DependencyModel/tests/AssemblyInfo.cs:8]: For numerous amounts of tests, we are throwing PNSE b/c of crypto.  

This method is the culprit:

https://github.com/dotnet/cli-migrate/blob/5395650fe2a118fb8d5e7d1475a0b684d969e908/src/Microsoft.DotNet.ProjectJsonMigration/Microsoft.DotNet.Internal.ProjectModel/DependencyContextBuilder.cs#L65-L87

File: src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs (hunk 3 above)
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs:12]: Do the types exercised by there tests actually work on Wasm Browser? E.g. can I add configuration file to my .csproj file and then get the file picked up by the default `IFileProvider` at runtime?

File: src/libraries/Microsoft.Extensions.DependencyModel/tests/AssemblyInfo.cs (hunk 4 above)
Comment (jkotas) [reply src/libraries/Microsoft.Extensions.DependencyModel/tests/AssemblyInfo.cs:8]: Thanks for the link

File: src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs (hunk 3 above)
Comment (steveisok) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs:12]: Good question.  If you look at the test results, we're likely at least somewhat functional.  

https://gist.github.com/steveisok/370622cdb0c41ff2bc3b439c730484e4

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems [hunk 5]
```diff
@@ -44,6 +44,8 @@
     <Compile Include="$(MSBuildThisFileDirectory)System\Activator.RuntimeType.cs" Condition="'$(TargetsCoreRT)' != 'true'" />
     <Compile Include="$(MSBuildThisFileDirectory)System\AggregateException.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\AppContext.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\AppContext.Browser.cs" Condition="'$(TargetsBrowser)' == 'true'" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\AppContext.UnixOrWindows.cs" Condition="'$(TargetsBrowser)' != 'true'" />
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:48]: Nit: I believe that we typically use `AnyOS.cs` for the generic implementation that works everywhere expect specific outliers.

File: src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs (hunk 3 above)
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/AppContext.Browser.cs:12]: I think we should have issues opened on making sure that the end-to-end works for these libraries. More passing tests does not necessarily mean that these libraries are actually usable by customers.

The linker analysis for single file unfriendly APIs (https://github.com/dotnet/runtime/issues/38405) should help us get a better handle on this.

If you believe that returning `/` here is going to help to make the affected scenarios work end-to-end, I can live with it.

File: src/libraries/Microsoft.Extensions.Configuration.Xml/tests/XmlConfigurationTest.cs [hunk 6]
```diff
@@ -432,7 +432,8 @@ public void XmlConfiguration_Does_Not_Throw_On_Optional_Configuration()
             var config = new ConfigurationBuilder().AddXmlFile("NotExistingConfig.xml", optional: true).Build();
         }
 
-        [ConditionalFact]
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/37669", TestPlatforms.Browser)]
         [FrameworkSkipCondition(RuntimeFrameworks.Mono)]
```
Comment (akoeplinger) [inline src/libraries/Microsoft.Extensions.Configuration.Xml/tests/XmlConfigurationTest.cs:437]: I think this can be removed as well, it doesn't do anything useful anymore since this is about the old .NET Framework-based Mono. We can do that in a separate PR, filed https://github.com/dotnet/runtime/issues/38878.

**Review by akoeplinger — APPROVED** [review]
